src/shared/math/position.ts

```diff
@@ -0,0 +1,218 @@
+const PRECISION_DECIMALS = 12;
+
+/**
+ * A slimmed-down representation for assets, restricted to what we need for math.
+ *
+ * We have an identifier for the kind of asset, which is needed to construct a position,
+ * and an exponent E, such that 10**E units of the base denom constitute a unit of the display denom.
+ *
+ * For example, 10**6 uUSD make up one USD.
+ */
+export interface Asset {
+  id: AssetId;
+  exponent: number;
+}
+
+/**
+ * A basic plan to create a position.
+ *
+ * This can then be passed to `planToPosition` to fill out the position.
+ */
+export interface PositionPlan {
+  baseAsset: Asset;
+  quoteAsset: Asset;
+  /** How much of the quote asset do you get for each unit of the base asset?
+   *
+   * This will be in terms of the *display* denoms, e.g. USD / UM.
+   */
+  price: number;
+  /** The fee, in [0, 10_000]*/
+  feeBps: number;
+  /** How much of the base asset we want to provide, in display units. */
+  baseReserves: number;
+  /** How much of the quote asset we want to provide, in display units. */
+  quoteReserves: number;
+}
+
+const priceToPQ = (
+  price: number,
+  pExponent: number,
+  qExponent: number,
+): { p: number; q: number } => {
+  // e.g. price     = X USD / UM
+  //      basePrice = Y uUM / uUSD = X USD / UM * uUSD / USD * UM / uUM
+  //                = X * 10 ** qExponent * 10 ** -pExponent
+  const basePrice = new BigNumber(price).times(new BigNumber(10).pow(qExponent - pExponent));
+  // USD / UM -> [USD, UM], with a given precision
+  const [q, p] = basePrice.toFraction(10 ** PRECISION_DECIMALS);
+  return { p: p.toNumber(), q: q.toNumber() };
+};
+
+const priceToPQ2 = (
+  price: number,
+  pExponent: number,
+  qExponent: number,
+): { p: Amount; q: Amount } => {
+  const pExponentUnits = BigInt(10) ** BigInt(pExponent);
+  const qExponentUnits = BigInt(10) ** BigInt(qExponent);
+
+  const scale = qExponentUnits < 1_000_000n ? 1_000_000n : 1n;
+
+  const p = pnum(
+    BigInt(
+      BigNumber((qExponentUnits * scale).toString())
+        .times(BigNumber(price))
+        .shiftedBy(-(qExponent - pExponent))
+        .toFixed(0),
+    ),
+    qExponent,
+  ).toAmount();
+
+  const q = pnum(pExponentUnits * scale, pExponent).toAmount();
+
+  return { p, q };
+};
+
+/**
+ * Convert a plan into a position.
+ *
+ * Try using `rangeLiquidityPositions` or `limitOrderPosition` instead, with this method existing
+ * as an escape hatch in case any of those use cases aren't sufficient.
+ */
+export const planToPosition = (plan: PositionPlan): Position => {
+  console.log('TCL: plan', plan);
+  // const { p, q } = priceToPQ(plan.price, plan.baseAsset.exponent, plan.quoteAsset.exponent);
+  const { p, q } = priceToPQ2(plan.price, plan.baseAsset.exponent, plan.quoteAsset.exponent);
+
+  // const r1 = pnum(plan.baseReserves, plan.baseAsset.exponent).toAmount();
+  // const r2 = pnum(plan.quoteReserves, plan.quoteAsset.exponent).toAmount();
+  const r1 = pnum(plan.baseReserves).toAmount();
```

Comment:
I think we need to make sure that base is asset_1 and quote is asset_2 and flip everything if it's not.

Ideally, there's a `Position` constructor that does this on your behalf so that you don't have to worry about it.